### ISSUE TITLE
Optimize JSON (De)serialization (Final Types et al)

### ISSUE THREAD

**Issue by dselman** [accordproject/concerto#482]
# Feature Request 🛍️

## Use Case

Allow modellers to declare that types are "final" and cannot be extended. This allows for optimisation of the serialised JSON, as an array of final types can no longer be polymorphic.

The $class is required in serialised instances because concepts can be extended. We could support the addition of a final keyword which would indicate that a concept could not be extended, meaning that the type of an object can be inferred by introspecting the model.

## Possible Solution

```
namespace acme

final concept Person {
   o String name
}

final concept Company {
   o String name
   o Person[] employees
}

const company = 
{
   "name" : "Clause",
   "employees" : [
      {"name" : "Dan"}, {"name", "Simon"}
   ]
}

// deserialize, passing in the name of the type explicitly
// during deserialization the model is introspected to know that company.employee is an Employee type (and cannot be a subclass).
serialiser.fromJSON( company, ''acme.Company' );
```

**Comment by dselman** [comment]
Note that if `$class` is missing we could assume that we need to do a type lookup - if the type found is abstract then we fail with an error.

**Comment by DS-AdamMilazzo** [comment]
I would suggest that the "final" feature is not needed for this, and you could get the desired savings in more situations via a more general mechanism. Since objects are naturally processed and validated recursively, you should already know the parent's type when you encounter a field value. From the parent's type definition, you already know the declared type of the field. $class could simply be omitted wherever it matches the declared field type. Also, when only the type name within the $class differs from the declared field type, the namespace and version could be omitted, leaving an abbreviated $class field containing only the type name. For example:

````
abstract concept Owner { ... }
concept HumanOwner extends Owner { ... }
concept CorporateOwner extends Owner { ... }

concept Thing {
  o Owner owner
}
````

The $class in the owner could simply be "HumanOwner" or "CorporateOwner". The namespace and version would be assumed to match that of the Owner type in that case.

**Comment by dselman** [comment]
> declared

Note that type determination based on the type of the property only works if you assume that the model doesn't change between when the instance was serialised and when it was deserialised. For example, if version 1 of the model is:

```
concept Owner { ... }

concept Thing {
  o Owner owner
}
```

_It is assumed that no $class is required, because the `owner` property is of an (non-abstract) type `Owner`._

Version 2 of the model is:

```
concept Owner { ... }
concept HumanOwner extends Owner { ... }
concept CorporateOwner extends Owner { ... }

concept Thing {
  o Owner owner
}
```

_A $class is now required as a type discriminator to deserialise the `Thing.owner` property_

The advantage of the `final` keyword (or sometimes called `sealed`) is that it makes it much clearer to the modeller than making a `final` type non-final is a breaking change - and that change occurs in the namespace of the type that is being declared "unfinal". Without  `final` keyword any introduction of a derived type, in any namespace, is potentially a breaking change for deserialisation.

**Comment by DS-AdamMilazzo** [comment]
> Note that type determination based on the type of the property only works if you assume that the model doesn't change between when the instance was serialised and when it was deserialised.

I don't think that's true. It should work regardless. The $class of the top-level object - which is always required - declares the exact version of the top-level object that you're dealing with. If it's v1 then you know you're in the first case. If it's v2 then you know you're in the second case. There's no ambiguity or assumptions about the model not changing.

(Furthermore, I don't see why a $class would be required even in your v2 model, at least according to my proposed rule. Given that Owner is not abstract, a $class-less owner would still be legal and unambiguous. It would only become required if Owner became abstract.)

**Comment by DS-AdamMilazzo** [comment]
It seems that your understanding of my proposal is that we allow $class to be omitted when and only when there's exactly one concrete type that a field could possibly be, so that adding more possible types suddenly makes $class required where it wasn't before. Something like this:
```
abstract concept A { }
concept C {
  o A a // $class can be omitted as long as there's only one subtype of A in existence anywhere???
}
```

I agree that would be a bad idea, but it's not my proposal. My proposal is much simpler and more generally applicable. It allows $class to be omitted _whenever a field type is not abstract_, and when omitted, it is understood to be the _exact_ concrete type the field is declared as. To revisit your v2 example above:

```
concept Owner { ... }
concept HumanOwner extends Owner { ... }
concept CorporateOwner extends Owner { ... }

concept Thing {
  o Owner owner
}
```

* Can $class be omitted for the "owner" field? Yes, because the field type, Owner, is not abstract.
* What type is inferred when $class is omitted? It is only the exact field type, M@2.0.0.Owner, and none other! If you want to use HumanOwner or CorporateOwner, then you need a $class.

I don't believe there is any form of model evolution for which this rule causes problems, assuming you are not mutating existing models without changing their versions, which can break all kinds of things. But if you have an example, I'd be glad to see it.

**Comment by DS-AdamMilazzo** [comment]
Return from whence thou came, bot.

**Comment by Mamatha1718** [comment]
I think the general mechanism is more versatile and practical than the $final keyword. It optimizes payloads dynamically without requiring explicit declarations, reduces complexity for modelers, and handles model evolution more gracefully. While $final might help in niche cases for strict type control, the general mechanism seems to achieve broader applicability with less risk of breaking changes. Would love to hear if there are specific scenarios where $final is essential!
Thank you.

**Comment by DS-AdamMilazzo** [comment]
I agree with everything you said. I don't think `$class` omission should be tied to a `final` keyword, because it can be handled generically.

That said, I do think a `final` keyword has value too. We've taken to annotating many of our concepts with a decorator to tell us that they're "sealed" (i.e. final), and we do other optimizations based on that, but of course it's not enforced by Concerto.

**Comment by Mamatha1718** [comment]
Thanks for explaining how the `final` keyword can be useful! I get now that it helps keep types fixed and allows for better performance by simplifying certain processes.

But I’m curious—if the general mechanism can already handle `$class` omission pretty well, do you think the `final` keyword is more of an extra feature for stricter design rules? Or do you feel it’s really necessary to unlock important optimizations for sealed types?

**Comment by Mamatha1718** [comment]
Got it, thank you for clarifying! It’s helpful to know that the optimizations for `final` types are specific to your use case and unrelated to `$class`. I agree that `$class` optimization is broadly valuable, and it’s great to see how `final` can add flexibility for other design purposes.
Thank You.

**Comment by TC5022** [comment]
@Mamatha1718 are you working on this issue? If not can this be assigned to me? @dselman
